plugins/modem-manager/fu-plugin-modem-manager.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <libmm-glib.h>
+
+#include "fu-plugin-vfuncs.h"
+
+#include "fu-mm-device.h"
+
+struct FuPluginData {
+	MMManager	*manager;
+};
+
+static void
+fu_plugin_mm_device_add (FuPlugin *plugin, MMObject *modem)
+{
+	const gchar *object_path = mm_object_get_path (modem);
+	g_autoptr(FuMmDevice) dev = NULL;
+	if (fu_plugin_cache_lookup (plugin, object_path) != NULL)
+		return;
+	dev = fu_mm_device_new (modem);
+	fu_plugin_device_add (plugin, FU_DEVICE (dev));
+	fu_plugin_cache_add (plugin, object_path, dev);
+}
+
+static void
+fu_plugin_mm_device_added_cb (MMManager *manager, MMObject *modem, FuPlugin *plugin)
+{
+	fu_plugin_mm_device_add (plugin, modem);
+}
+
+static void
+fu_plugin_mm_device_removed_cb (MMManager *manager, MMObject *modem, FuPlugin *plugin)
+{
+	const gchar *object_path = mm_object_get_path (modem);
+	FuMmDevice *dev = fu_plugin_cache_lookup (plugin, object_path);
+	if (dev == NULL)
+		return;
+	g_debug ("removed modem: %s", mm_object_get_path (modem));
+	fu_plugin_cache_remove (plugin, object_path);
+	fu_plugin_device_remove (plugin, FU_DEVICE (dev));
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	g_autolist(GDBusObject) modems = NULL;
+
+	modems = g_dbus_object_manager_get_objects (G_DBUS_OBJECT_MANAGER (priv->manager));
+	for (GList *l = modems; l != NULL; l = g_list_next (l)) {
+		MMObject *modem = MM_OBJECT (l->data);
+		fu_plugin_mm_device_add (plugin, modem);
+	}
+	return TRUE;
+}
+
+gboolean
+fu_plugin_startup (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	g_autoptr(GDBusConnection) connection = NULL;
+
+	connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, error);
+	if (connection == NULL)
+		return FALSE;
+	priv->manager = mm_manager_new_sync (connection,
```

Comment:
Have you experimented at all what exactly happens if the modem manager daemon crashes (and/or comes back)?  I would wonder if some special handling will be needed for that particular scenario around device removal and this plugin.